### ISSUE TITLE
is it possible to test array against array

### ISSUE THREAD

**Issue by agonencl** [jruizgit/rules#100]
is it possible to compare with two arrays, I'm trying to merge facts based on array overlap?
for example:
```
from durable.lang import *
import uuid

with ruleset('aa'):
	@when_all( +m.ip & -m.incident_id)
	def new_event(c):
		print ('creating incident from event {}'.format(c.m.ip))
		c.assert_fact({'incident_id': str(uuid.uuid4()),'ip': [c.m.ip]})

	@when_all(c.first << (+m.incident_id),
			  c.second << (m.incident_id <> c.first.incident_id & 
						   m.ip.anyItem(item in c.first.ip))
	def merge_list(c):
		# retract both first and second facts
		# put a merge fact
	
	@when_all(+s.exception)
	def second(c):
		print(c.s.exception)
		c.s.exception = None

run_all()
```

**Comment by jruizgit** [comment]
Hi, thanks a lot for asking the question. durable_rules doesn't support correlating events based on array values. Right now arrays can only be filtered by Alpha nodes. The 'in' operator is not supported. I'm planning to add support for such features within the next few weeks. You will be able to write an expression like this:

```python
@when_all(c.first << (+m.incident_id),
              c.second << ((m.incident_id != c.first.incident_id) & m.ip.anyItem(item in c.first.ip)))
    def merge_list(c):
        print('Hello')
```

**Comment by agonencl** [comment]
Thanks, this will be a great feature, currently, I'm blocked with my testing since my logic depends on merging incidents. 
btw , can you explain what does it mean Alpha & Beta nodes ? is there some documention on it?

**Comment by agonen** [comment]
any update on this ?

**Comment by jruizgit** [comment]
Hi, I haven't finished the feature implementation yet. When a ruleset is started, durable_rules creates a Rete network with Alpha and Beta nodes. Alpha nodes are simple filters, while Beta nodes keep inference state. durable_rules stores and manages Beta node data in redis. This blog post explains the ideas in a little more detail: https://jruizblog.com/2015/02/23/rete_d/

**Comment by agonencl** [comment]
ETA ?

**Comment by jruizgit** [comment]
I have pushed the first part of the implementation (also published to pypi). At this point the simple scenario works (example below). It will take me one more week to finish the more advanced scenarios.

```python
from durable.lang import *

with ruleset('risk'):

    @when_all(c.first << +m.max,
                       c.second << m.payments.anyItem(item > c.first.max)
    def rule1(c):
        print('fraud 1 detected {0}'.format(c.second.payments))

    @when_start
    def start(host):
        host.assert('risk', {'max': 400})
        host.post('risk', {'payments': [ 150, 300, 450 ]})
```

**Comment by agonen** [comment]
Thanks. Great achievement. Can I assume that not exists operator like "! =" will work as well? Do you plan to set  operations on arrays. Like if array is in another array?

**Comment by jruizgit** [comment]
Yes other operators such as `!=`, `==`, `<`, `<=`, `>`, `>=` should work. Both anyItem and allItems functions should work too. Arrays of objects and arrays of arrays don't work yet (I'm planning to add support for them this week). 

I haven't implemented any support for comparing two arrays. Once I complete the work above it should be relatively easy to add.

**Comment by hassanzade** [comment]
Hi, is there any update on the 'in' operator like m.ip.anyItem(item in c.first.ip) ? I am trying to compare two lists and I don't know how I can do it. Thanks.
